examples/scripts/TreeCompareAuto.C

```diff
@@ -0,0 +1,293 @@
+// fails.
+
+#include <RtypesCore.h>
+#include <TBranch.h>
+#include <TFile.h>
+#include <TH1.h>
+#include <TMath.h>
+#include <TObjArray.h>
+#include <TString.h>
+#include <TTree.h>
+#include <iostream>
+#include <sstream>
+#include <string>
+#include <utility>
+#include <vector>
+
+std::vector<TString> GetLeafNames(TTree&);
+
+bool operator==(TH1 const&, TH1 const&);
+
+int TreeCompareAuto(TString fileName1 = "", TString fileName2 = "")
+{
+    if (fileName1.IsNull() || fileName2.IsNull()) {
+        cout << "Filenames are not defined" << endl;
+        exit(42);
+    }
+
+    // Get the output tree from the original file
+    std::unique_ptr<TFile> file1{TFile::Open(fileName1, "READ")};
+    if (!file1) {
+        std::cout << "Could not open file " << fileName1 << std::endl;
+        return 42;
+    }
+    std::unique_ptr<TTree> tree1{file1->Get<TTree>("cbmsim")};
+    if (!tree1) {
+        std::cout << "File " << fileName1 << " does not have the cbmsim tree." << std::endl;
+        return 42;
+    }
+
+    // Get the output tree from the file which should be compared
+    std::unique_ptr<TFile> file2{TFile::Open(fileName2, "READ")};
+    if (!file2) {
+        std::cout << "Could not open file " << fileName2 << std::endl;
+        return 42;
+    }
+    std::unique_ptr<TTree> tree2{file2->Get<TTree>("cbmsim")};
+    if (!tree2) {
+        std::cout << "File " << fileName2 << " does not have the cbmsim tree." << std::endl;
+        return 42;
+    }
+
+    // Add the output tree from the file to compare as friend to the tree
+    // of the original file. This allows to access a data member of the
+    // original file by e.g. StsHit.fX and the data element of the second tree
+    // by tree2.StsHit.fX
+    tree1->AddFriend(tree2.get(), "tree2");
+
+    // Get the leaf names for all leaves which should be compared
+    // from the first input tree
+    auto leaves = GetLeafNames(*tree1.get());
+
+    if (0 == leaves.size()) {
+        std::cout << "Test passed. Tree does not contain nay branches which must be checked" << std::endl;
+        return 0;
+    }
+
+    TCanvas c1;
+
+    std::stringstream outstream;
+    bool okay{true};
+    int numTestedLeaves{0};
+    int numEmptyLeaves{0};
+    int numLeaves{0};
+    int numFailedLeaves{0};
+    int numIdenticalEntries{0};
+    int numIdenticalHistograms{0};
+    int numKolmogorovHistograms{0};
+
+    for (auto leaf : leaves) {
+        TString leafName = leaf;
+        TString leafName1 = "tree2." + leafName;
+
+        TString command1 = leafName + ">>htemp";
+        tree1->Draw(command1);
+        int entries{0};
+        float low{0.};
+        float high{0.};
+        int nBins{0};
+        auto htemp = static_cast<TH1F*>(gPad->GetPrimitive("htemp"));
+        if (htemp) {
+            entries = htemp->GetEntries();
+            nBins = htemp->GetNbinsX();
+            low = htemp->GetXaxis()->GetXmin();
+            high = htemp->GetXaxis()->GetXmax();
+        }
```

Comment:
```suggestion
        TH1F htemp;
        long long entries{tree1->GetEntries()};
        float value = 0;
        tree1->SetBranchAddress(leafName, &value);
        for (long long i = 0; i < entries; ++i) {
            tree1->GetEntry(i);
            htemp->Fill(value);
        }
        float low = htemp.GetXaxis()->GetXmin();
        float high = htemp.GetXaxis()->GetXmax();
        int nBins = htemp.GetNbinsX();
```

Why not something simple like this (have not tested the above code, but there should be some code similar to the above)?

What is the benefit to go through all this `Draw()` logic and global referencing via `gPad`?

Reply:
For the `hist` case a bit further down, one could use a `TFormula`, but since it is just a subtraction, not even needed?

Reply:
I think this will not work since the value can be anything and must not be a float.

Reply:
Then why do you cast it to a `TH1F*` where the `F` stands for `float`? (Also, how are `low` and `high` then supposed to work if it is not a float?)

Reply:
> (Also, how are low and high then supposed to work if it is not a float?)

Btw, [`TAxis::GetXmax()`](https://root.cern.ch/doc/v624/classTAxis.html#a5ae359f6f1fbbb078cd425a383582b11) returns a `double`?!

Reply:
Also check [`Hist.Precision.1D`](https://github.com/root-project/root/blob/f83a40912dbae9b42e7df1a89774a87ebbc7a8fb/README/ReleaseNotes/v610/index.md?plain=1#L124-L131). If I understand this correctly, `TTree::Draw()` will always produce single precision histograms, unless you change it globally in your ROOT setting?!

Reply:
In our case, it should rather depend on the data in the root file, should it not?

Reply:
You can pass any number to TH1F::Fill() which will find the proper bin to increase. The type of the histogram only defines the maximum value which can be stored per bin. TH1C can only store values less than 128 per bin. TH1F stores the bin content in a float value.

These are the values you have to pass to the constructor which is the same for all types of one dimensional histograms

[in] | name | name of histogram (avoid blanks) | const char*
[in] | title | histogram title. | const char*
[in] | nbins | number of bins | int
[in] | xlow | low edge of first bin | double
[in] | xup | upper edge of last bin (not included in last bin) | double

TAxis::GetMax returns the value value of the upper edge of the last bin which is a double.

Reply:
Ok, but then the common type to fill the histogram with is just `double`? So, read the tree content into doubles?

Reply:
Or as a question: Which `H1F` API is used by `TTree::Draw` to supply add its stored values to the histogram?

Reply:
And whatever the code looks like, my points are:
1. Why do we need to involve a `TCanvas` object, a `TPad` object and a function called `Draw` - all clearly graphics related entities - in a script that does not want to produce any graphics? Especially, if we have to create a random `TCanvas` object in the middle of it to silence warnings? Don't you agree that this is a terrible program to write, read and maintain?
2. Can we produce the histogram object not through the indirection of the global object `gPad`?